### PR TITLE
[Bug fix]: Qwen2.5-vl use eagle3 in spec v2

### PR DESCRIPTION
**Pull request by jiapingW**
## Motivation
Fix the bug: https://github.com/sgl-project/sglang/issues/14060. And combined impl with https://github.com/sgl-project/sglang/pull/13904.
TODO:  Align qwen2.5-vl accept length in spec v2 with spec v1.
Spec V1:
``` 
{
    "mmstar": [
        {
            "batch_size": 1,
            "steps": null,
            "topk": null,
            "num_draft_tokens": null,
            "metrics": [
                {
                    "latency": 33.99738936708309,
                    "output_throughput": 198.132860357858,
                    "accept_length": 3.1803588290840414,
                    "accuracy": 0.33,
                    "num_questions": 100,
                    "num_valid_predictions": 86,
                    "categorical_performance": null
                }
            ],
            "num_samples": 100
        }
    ]
}
```
Spec V2:
```
{
    "mmstar": [
        {
            "batch_size": 1,
            "steps": null,
            "topk": null,
            "num_draft_tokens": null,
            "metrics": [
                {
                    "latency": 49.90318745514378,
                    "output_throughput": 147.0246766620864,
                    "accept_length": 2.1396908719743366,
                    "accuracy": 0.32,
                    "num_questions": 100,
                    "num_valid_predictions": 76,
                    "categorical_performance": null
                }
            ],
            "num_samples": 100
        }
    ]
}
```

## Accuracy Tests

Test use mmstar bencmark:
```
[2025-11-27 23:10:52] INFO:     127.0.0.1:53272 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:52] Prefill batch, #new-seq: 1, #new-token: 233, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:52] Decode batch, #running-req: 1, #token: 315, token usage: 0.00, accept len: 2.38, accept rate: 0.10, cuda graph: True, gen throughput (token/s): 147.79, #queue-req: 0, 
[2025-11-27 23:10:53] Decode batch, #running-req: 1, #token: 384, token usage: 0.00, accept len: 1.75, accept rate: 0.07, cuda graph: True, gen throughput (token/s): 152.98, #queue-req: 0, 
[2025-11-27 23:10:53] INFO:     127.0.0.1:53280 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:53] Prefill batch, #new-seq: 1, #new-token: 233, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:53] INFO:     127.0.0.1:53292 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:53] Prefill batch, #new-seq: 1, #new-token: 270, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:53] Decode batch, #running-req: 1, #token: 337, token usage: 0.00, accept len: 2.15, accept rate: 0.09, cuda graph: True, gen throughput (token/s): 130.64, #queue-req: 0, 
[2025-11-27 23:10:54] INFO:     127.0.0.1:53306 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:54] Prefill batch, #new-seq: 1, #new-token: 230, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:54] Decode batch, #running-req: 1, #token: 315, token usage: 0.00, accept len: 2.15, accept rate: 0.09, cuda graph: True, gen throughput (token/s): 155.84, #queue-req: 0, 
[2025-11-27 23:10:54] INFO:     127.0.0.1:53312 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:54] Prefill batch, #new-seq: 1, #new-token: 236, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:55] Decode batch, #running-req: 1, #token: 0, token usage: 0.00, accept len: 2.08, accept rate: 0.09, cuda graph: True, gen throughput (token/s): 150.25, #queue-req: 0, 
[2025-11-27 23:10:55] INFO:     127.0.0.1:53320 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:55] Prefill batch, #new-seq: 1, #new-token: 233, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:55] Decode batch, #running-req: 1, #token: 371, token usage: 0.00, accept len: 2.10, accept rate: 0.09, cuda graph: True, gen throughput (token/s): 152.71, #queue-req: 0, 
[2025-11-27 23:10:55] INFO:     127.0.0.1:53332 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:55] Prefill batch, #new-seq: 1, #new-token: 215, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:56] Decode batch, #running-req: 1, #token: 299, token usage: 0.00, accept len: 1.98, accept rate: 0.08, cuda graph: True, gen throughput (token/s): 145.16, #queue-req: 0, 
[2025-11-27 23:10:56] INFO:     127.0.0.1:53348 - "POST /generate HTTP/1.1" 200 OK
[2025-11-27 23:10:56] Prefill batch, #new-seq: 1, #new-token: 267, #cached-token: 14, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-11-27 23:10:56] Decode batch, #running-req: 1, #token: 356, token usage: 0.00, accept len: 2.00, accept rate: 0.08, cuda graph: True, gen throughput (token/s): 142.13, #queue-req: 0, 
[2025-11-27 23:10:57] Decode batch, #running-req: 1, #token: 441, token usage: 0.00, accept len: 2.12, accept rate: 0.09, cuda graph: True, gen throughput (token/s): 185.65, #queue-req: 0, 
```
## Checklist

- [✅] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [✅] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

**Comment by Terry-Uv** [comment]
Hi,

I reproduced the problem in #14060, and with the current changes in this PR, the error has disappeared.

To inspect the behavior further, I ran a small controlled test on this branch with `Qwen2.5-VL-7B-Instruct` + `EAGLE3`:

- **Request**

  - OpenAI-compatible `/v1/chat/completions` with a single image and the prompt:  “用一句话描述这张图片。”
  - Sampling parameters: `temperature = 0.0`, `top_p = 1.0`, `seed = 42`, `max_tokens = 128`.

- **Spec V1** (no `SGLANG_ENABLE_SPEC_V2`):

  - Multiple runs give a stable output:

    > “这张图片展示了一位女士和她的狗狗在海滩上愉快互动的温馨场景。”

  - `/get_server_info` shows `disable_overlap_schedule = true`.

- **Spec V2** (`SGLANG_ENABLE_SPEC_V2=1`):

  - Multiple runs also give a stable (but different) output:

    > “这张图片展示!!女士和她的狗狗!!在沙滩!上，!日落时光!!!”

  - `/get_server_info` shows `disable_overlap_schedule = false` (overlap spec enabled).

So for a fixed prompt and sampling config, both Spec V1 and Spec V2 behave deterministically within each mode, and the original `'list' object has no attribute 'to'` error is gone.

Though a little strange, the fact that Spec V1 / Spec V2 produce different responses under the same seed still seems acceptable to me, since the sampling calls (and overlap scheduling) differ between the two pipelines?

For the MMStar benchmark TODO (Spec V1 `accept_length` ~ 3.18 vs Spec V2 ~ 2.14),  
could you share the exact `run_mmstar.py` + command you used?  I couldn’t find it in this repo :(
I’d like to try a smaller run locally to see how things look like.

**Comment by Terry-Uv** [comment]
Though having some trouble accessing dataset at first, I still managed to run the mmstar benchmark. I read the benchmark part within SpecForge, make sure the validate dataset remains the same, and get the results below:

**Spec V1:**

```json
{
   "mmstar": [
       {
           "batch_size": 1,
           "steps": null,
           "topk": null,
           "num_draft_tokens": null,
           "metrics": [
               {
                   "latency": 58.92725312989205,
                   "output_throughput": 112.1382662353891,
                   "accept_length": 3.507430997876858,
                   "accuracy": 0.33,
                   "num_questions": 100,
                   "num_valid_predictions": 83,
                   "categorical_performance": null
               }
           ],
           "num_samples": 100
       }
   ]
}
```

**Spec V2:**

```json
{
   "mmstar": [
       {
           "batch_size": 1,
           "steps": null,
           "topk": null,
           "num_draft_tokens": null,
           "metrics": [
               {
                   "latency": 89.52508324664086,
                   "output_throughput": 67.26606423151196,
                   "accept_length": 2.037901861252115,
                   "accuracy": 0.31,
                   "num_questions": 100,
                   "num_valid_predictions": 74,
                   "categorical_performance": null
               }
           ],
           "num_samples": 100
       }
   ]
}
```


-  Since `accuracy` and `num_valid_predictions` seems close between Spec V1 and Spec V2,  the Spec V2 path looks functionally correct at first sight.
- However, `accept_length` drops from ~3.5 to ~2.0, and latency / throughput regress by roughly the same factor. This matches the TODO.

My current guess is that this remaining gap is not caused by the Qwen2.5-VL-specific changes in this PR, but rather by the EAGLE3 + overlap integration in Spec V2 itself, e.g. :

* how `EagleVerifyInput.sample()` in `eagle_info_v2.py` maps draft tokens (`draft_token`, `positions`, `retrive_*`) back to target logits under the overlap scheduler, and
* possibly some subtle interaction with multimodal `mrope_positions` / `seq_lens` in the decode path for Qwen2.5-VL.

If this direction sounds reasonable, I’d be happy to dig a bit deeper, like just beginning with the verification logic:

1. On a tiny MMStar subset, log per-request `accept_length` / tree topology and compare Spec V1 vs Spec V2 under the same sampling config.
2. Double-check that the Spec V2 verification path (`eagle_info_v2.sample` + `scheduler_output_processor_mixin`) accounts for accepted tokens in exactly the same way as the Spec V1 path.

Please let me know if this matches your expectations for Spec V2 on Qwen2.5-VL, or if there is ongoing work on EAGLE3 + overlap / MMStar that I should align with before touching the verification logic.

If there're any other possible paths you want me to inspect (maybe suspicious algo logic?), please inform me too. I'd be happy to take them.

Thanks again for letting me co-work on this! 😀

**Comment by hnyls2002** [comment]
Hi @jiapingW, thanks for the fix. The qwen2.5-vl eagle3 pieces have since landed on main separately: the aux_hidden_states pass-through merged via #13904, and the mrope computation now covers the spec-v2 forward modes (TARGET_VERIFY / DRAFT_EXTEND_V2) in forward_batch_info. The motivating issue #14060 was closed as completed. Closing as superseded - please reopen if I've missed something.
